File: ABM-SIM-DOTNET/src/taker/TFundamentalist.cs
namespace ABMSIM {
    public class TFundamentalist : IAgent {
        private int fundamentalistPrice;
        public TFundamentalist(int fundamentalistPrice) : base() {
            this.fundamentalistPrice = fundamentalistPrice;
        }

        private float getVolumeConstant(float midPrice) {
            if (Math.Abs(fundamentalistPrice - midPrice) <= (midPrice) * (2f)) {
                return 20f;
            } else if (Math.Abs(fundamentalistPrice - midPrice) > (midPrice) * (2f)) {
                return 50f;
            } else {
                return 0f;
            }
        }

        public MarketOrder? Action(Moment m) {
            OrderTypes orderType;
            if (fundamentalistPrice < m.midPrice - (0.5f) * (m.spread)) {
                orderType = OrderTypes.BUY;
            } else if (fundamentalistPrice > m.midPrice + (0.5f) * (m.spread)) {
                orderType = OrderTypes.SELL;
            } else {
                return null;
            }
            int volume = (int)Util.calculatePowerLawVolume(orderType == OrderTypes.BUY, getVolumeConstant(m.midPrice), m);
            return new MarketOrder(orderType, volume, fundamentalistPrice);
        }
    }
}

File: abm-sim-js/src/sim/agents/taker/TFundamentalist.js
import MarketOrder, { OrderType } from "../../MarketOrder.js"
import {calculatePowerLawVolume} from '../../Util.js'

export default class TFundamentalist {
  fundamentalistPrice;
  constructor(fundamentalistPrice) {
    this.fundamentalistPrice = fundamentalistPrice
  }

  getVolumeConstant(midPrice) {
    if (Math.abs(this.fundamentalistPrice - midPrice) <= (midPrice) * 2) {
      return 20;
    } else if (Math.abs(this.fundamentalistPrice - midPrice) > (midPrice) * 2) {
      return 50
    } else {
      return 0
    }
  }

  Action(moment) {
    let orderType
    if (this.fundamentalistPrice < moment.midPrice - (0.5 * moment.spread)) {
      orderType = OrderType.BUY
    } else if (this.fundamentalistPrice > moment.midPrice - (0.5 * moment.spread)) {
      orderType = OrderType.SELL
    } else {
      return null
    }
    const volume = calculatePowerLawVolume(orderType === OrderType.BUY, this.getVolumeConstant(moment.midPrice), moment)
    return new MarketOrder(orderType, Math.floor(volume), moment.midPrice - (0.5 * moment.spread))
  }
}